### PR TITLE
dracut: add EFI kernel hook

### PR DESCRIPTION
**Pull request by ericonr**
Also made some small formatting changes.

@ahesford

### CODE REVIEW

**Comment by ericonr** [comment]
@sgn could you take a look at this one?

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+if [ "x${CREATE_UEFI_BUNDLES}" != x1 ]; then
+	exit 0
+fi
+
+if [ ! -x bin/dracut ]; then
+	exit 0
+	fi
+
+if [ -z "${KERNEL_CMDLINE}" ] && [ "x${SKIP_KERNEL_CMDLINE}" != x1 ]; then
+	echo "ERROR: You need to set the kernel command line in /etc/default/dracut-uefi-hook"
```
Comment (Duncaen) [inline srcpkgs/dracut/files/kernel-uefi-hook-postinst:20]: Is this logic necessary?

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+	exit 1
+fi
+
+OPTIONS="${OPTIONS} --force --uefi ${UEFI_BUNDLE_DIR:=boot/efi/EFI/Void}/linux-${VERSION}.efi ${VERSION}"
```
Comment (Duncaen) [inline srcpkgs/dracut/files/kernel-uefi-hook-postinst:24]: `${foo:-default}` as you don't need to assign it.

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+#
+# Kernel post-install hook for dracut EFI bundle.
+#
+# Arguments passed to this script: $1 pkgname, $2 version.
+#
+PKGNAME="$1"
+VERSION="$2"
+
+. "${ROOTDIR}/etc/default/dracut-uefi-hook"
+if [ "x${CREATE_UEFI_BUNDLES}" != x1 ]; then
```
Comment (Duncaen) [inline srcpkgs/dracut/files/kernel-uefi-hook-postinst:11]: I usually prefer set/unset for things like that.
and then just have something 
```
# uncomment to enable...
#CREATE_UEFI_BUNDLES=yes
```

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst (hunk 2 above)
[the remark this answers is not in the document]
Comment (sgn) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:24]: And I'm thinking about supporting this in `sbsigntool` and `refind` hooks.

File: srcpkgs/dracut/files/dracut-uefi-hook.confd [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# Options for the dracut-uefi kernel hook script installed by the dracut package.
+
+CREATE_UEFI_BUNDLES=0
+# To allow dracut to create EFI bundles, set
+#CREATE_UEFI_BUNDLES=1
+# This requires an EFI stub, usually provided by the gummiboot package,
```
[the remark this answers is not in the document]
Comment (sgn) [reply srcpkgs/dracut/files/dracut-uefi-hook.confd:6]: disregard this one, I was wrong.

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+	fi
+
+if [ -z "${KERNEL_CMDLINE}" ] && [ "x${SKIP_KERNEL_CMDLINE}" != x1 ]; then
+	echo "ERROR: You need to set the kernel command line in /etc/default/dracut-uefi-hook"
+	exit 1
+fi
+
+OPTIONS="${OPTIONS} --force --uefi ${UEFI_BUNDLE_DIR:=boot/efi/EFI/Void}/linux-${VERSION}.efi ${VERSION}"
+
+dracut --kernel-cmdline="${KERNEL_CMDLINE}" ${OPTIONS}
```
Comment (sgn) [inline srcpkgs/dracut/files/kernel-uefi-hook-postinst:26]: `dracut` is talkative, `-q` should be used, I think.

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst (hunk 1 above)
Comment (ericonr) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:20]: It's solving the issue that if dracut doesn't find a `kernel_cmdline` entry in a `dracut.conf` file and isn't passed a cmdline in the command line argumets, it will prompt the user for a cmdline.

Comment (ericonr) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:20]: I couldn't think of another way of solving this particular issue.

File: srcpkgs/dracut/files/dracut-uefi-hook.confd [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+# as well as setting the kernel command line.
+
+# You can set the kernel command line with
+#KERNEL_CMDLINE="root=..."
```
[the remark this answers is not in the document]
Comment (ericonr) [reply srcpkgs/dracut/files/dracut-uefi-hook.confd:10]: I picked the one from our Grub config.

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst (hunk 2 above)
[the remark this answers is not in the document]
Comment (ericonr) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:24]: > And I'm thinking about supporting this in sbsigntool and refind hooks.

How would this look?

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst (hunk 5 above)
Comment (ericonr) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:26]: Yup!

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst (hunk 1 above)
Comment (ericonr) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:20]: The other way of solving this would be always passing at least an empty string to the command line parameter.

Comment (ahesford) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:20]: That seems like a good option, provided it has no unanticipated side effects.

Comment (ericonr) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:20]: I would still prefer erroring out, but if you think it's fine.

Comment (ericonr) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:20]: Hmm there's also the use case without secure boot that allows the bundle to use the cmdline from refind/grub, so the command line could be empty.

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+if [ -z "${KERNEL_CMDLINE}" ] && [ "x${SKIP_KERNEL_CMDLINE}" != x1 ]; then
+	echo "ERROR: You need to set the kernel command line in /etc/default/dracut-uefi-hook"
+	exit 1
+fi
+
+DRACUT_OPTIONS="${DRACUT_OPTIONS} --force --uefi ${UEFI_BUNDLE_DIR:=boot/efi/EFI/void}/linux-${VERSION}.efi ${VERSION}"
+mkdir -p "${UEFI_BUNDLE_DIR}"
+
+dracut -q --kernel-cmdline="${KERNEL_CMDLINE}" ${DRACUT_OPTIONS}
```
Comment (sgn) [inline srcpkgs/dracut/files/kernel-uefi-hook-postinst:27]:
```sh
dracut -q ${KERNEL_CMDLINE:+--kernel-cmdline="${KERNEL_CMDLINE}"} ${DRACUT_OPTIONS}
```
looks better?

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst (hunk 1 above)
Comment (sgn) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:20]: Just write a warning in the configuration file, and let the user see the boot failure :shrug: 
Don't bother with it, I would say.

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst (hunk 2 above)
[the remark this answers is not in the document]
Comment (sgn) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst]: Force them read this hooks if exists, maybe.
Rather messy but I don't like different parts of system refusing co-operating with each others. Just make sure variable introduced by this change doesn't overlap with them, I can make it works.
I like this idea, (so I can sign the whole efi, my initrd is unsigned, now) but I don't use gummiboot, for example.

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst (hunk 7 above)
Comment (ahesford) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:27]: I think the issue is that, if you don't set `kernel_cmdline` in the config file or the `dracut` command line, the user will be prompted for a value, which will stall the hook. @ericonr, is this correct?

Comment (ericonr) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:27]: It does!

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst (hunk 2 above)
[the remark this answers is not in the document]
Comment (sgn) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst]: Something like this to prepare for it
https://github.com/sgn/void-packages/commit/9bbcb81662861205e5fb70ceeb567d2b1a56a281

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst (hunk 7 above)
Comment (ericonr) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst:27]: Gah I hadn't seen @ahesford's comment.

So, that one does look better, and if I add the `-q` argument to dracut it simply plows ahead and gets an empty cmdline. So it should be fine.

File: srcpkgs/dracut/files/kernel-uefi-hook-postinst (hunk 2 above)
[the remark this answers is not in the document]
Comment (ericonr) [reply srcpkgs/dracut/files/kernel-uefi-hook-postinst]: Cool! I like the idea.

**Comment by ericonr** [comment]
Updated again.

**Comment by sgn** [comment]
Is this ready?

Or are we better to split the uefi bundle hook and gummiboot?

If this is ready, I'll merge this in, without worrying about other option.

**Comment by ericonr** [comment]
I don't think it's fully ready, the post remove hook isn't using `-z` to detect if it should do anything. Btw, how do we deal with 

```
if [ ! -x bin/dracut ]; then
	exit 0
fi
```

? Should it exit with 1 and show an error, since having the hook installed but not dracut is an issue?

File: srcpkgs/dracut/files/kernel-uefi-hook-postrm [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+#
+# Kernel post-remove hook for dracut EFI bundle.
+#
+# Arguments passed to this script: $1 pkgname, $2 version.
+#
+PKGNAME="$1"
+VERSION="$2"
+
+. "${ROOTDIR}/etc/default/dracut-uefi-hook"
+if [ "x${CREATE_UEFI_BUNDLES}" != x1 ]; then
```
Comment (sgn) [inline srcpkgs/dracut/files/kernel-uefi-hook-postrm:11]: I don't think we need this check.
Just check if the file exist then remove it?

File: srcpkgs/dracut/files/kernel-uefi-hook-postrm [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+	exit 0
+fi
+
+: "${UEFI_BUNDLE_DIR:=boot/efi/EFI/Void}"
```
Comment (sgn) [inline srcpkgs/dracut/files/kernel-uefi-hook-postrm:15]: Lowercase `V` in `Void`.
Since grub will create this very same directory with lowercase `void`.
I'm not sure what they will do if they're conflicting now.
And I couldn't say about their future.

File: srcpkgs/dracut/files/kernel-uefi-hook-postrm (hunk 8 above)
Comment (ericonr) [reply srcpkgs/dracut/files/kernel-uefi-hook-postrm]: Agree.

**Comment by sgn** [comment]
> I don't think it's fully ready, the post remove hook isn't using `-z` to detect if it should do anything.
> Btw, how do we deal with
> 
> ```
> if [ ! -x bin/dracut ]; then
> 	exit 0
> fi
> ```

If this hook is included with `dracut`, this check is likely unnecessary.
Albeit, we should use `usr/bin/dracut` instead?

But, this hook is depended on both `dracut` and an efi file provided by `gummiboot`.
I'm not sure if it's better to split it out and explicitly `depends` on `gummiboot`?

> ? Should it exit with 1 and show an error, since having the hook installed but not dracut is an issue?

**I** don't care about the broken system _if their owners choose to break it_

**Comment by ericonr** [comment]
> If this hook is included with dracut, this check is likely unnecessary.

I agree. It is mostly following the style of the other hook, that's why it has the check.

> I'm not sure if it's better to split it out and explicitly depends on gummiboot?

Hmm that might be a better way of doing it. We mention the dependency in the config file, at least. Would it be okay to have a subpackage only for the hook and the conf file?

**Comment by sgn** [comment]
On 2020-07-27 05:59:09-0700, Érico Nogueira Rolim <notifications@github.com> wrote:
> > If this hook is included with dracut, this check is likely unnecessary.
> 
> I agree. It is mostly following the style of the other hook, that's why it has the check.
> 
> > I'm not sure if it's better to split it out and explicitly depends on gummiboot?
> 
> Hmm that might be a better way of doing it. We mention the
> dependency in the config file, at least. Would it be okay to have
> a subpackage only for the hook and the conf file?

I support for subpackage, don't know about other.
We have a ton of meta package already,
and despite this package is a single POSIX script and a config,
it must be installed in system level.

Still better than `sx` which is single 4x line bash script,
which I want to remove badly.

-- 
Danh

**Comment by ericonr** [comment]
Heh, ok! Will split it off.

**Comment by ericonr** [comment]
Done.

**Comment by sgn** [comment]
Looks good. I hope you don't mind to wait for few days to let other people complain :D

**Comment by ericonr** [comment]
No worries, I'm not even using it any more :p 

Thanks!

File: srcpkgs/dracut/template [hunk 10]
```diff
@@ -14,17 +14,25 @@ license="GPL-2.0-or-later, LGPL-2.0-or-later"
 homepage="http://www.kernel.org/pub/linux/utils/boot/dracut/dracut.html"
 distfiles="https://github.com/dracutdevs/dracut/archive/${version}.tar.gz"
 checksum=f9dbf18597e5929221365964293212c8c9ffb7d84529c5a338c834ecab06e333
+subpackages="dracut-network"
```
Comment (ahesford) [inline srcpkgs/dracut/template:17]: Why is this needed, and should dracut-uefi be included?

Comment (ahesford) [reply srcpkgs/dracut/template:17]: Nevermind, didn't read far enough down...

**Comment by ericonr** [comment]
Added binutils to dracut-uefi, it needs it for the `objcopy` step.

File: srcpkgs/dracut/template [hunk 11]
```diff
@@ -14,17 +14,25 @@ license="GPL-2.0-or-later, LGPL-2.0-or-later"
...
 case "$XBPS_TARGET_MACHINE" in
 *-musl) # Add musl-fts implementation
 	makedepends+=" musl-fts-devel"
 	;;
 esac
 
+case "$XBPS_TARGET_MACHINE" in
+	i686*|x86_64*|aarch64*) # EFI bundle config and hooks
+		_efi_bundle=1
+		subpackages+=" dracut-uefi"
```
Comment (sgn) [inline srcpkgs/dracut/template:32]: `uefi-bundle` is not supported in aarch64. (`/usr/bin/dracut:1140`)

Comment (ericonr) [reply srcpkgs/dracut/template:32]: Oh, good find. Wonder if they forgot to add aarch64 or if it doesn't work.

Comment (sgn) [reply srcpkgs/dracut/template]: I guess it's ready, I'll pull this patch in https://github.com/dracutdevs/dracut/pull/871/commits/6184c7a9fed3dd75669e2545a4024ed690d9336e

File: srcpkgs/dracut/template [hunk 12]
```diff
@@ -52,8 +67,19 @@ dracut-network_package() {
...
 		for f in 40network 90livenet 95fcoe 95iscsi \
-		    95nbd 95nfs 95zfcp 95znet; do
+			95nbd 95nfs 95zfcp 95znet; do
 			vmove usr/lib/dracut/modules.d/${f}
 		done
 	}
 }
+
+dracut-uefi_package() {
+	depends="binutils gummiboot dracut-${version}_${revision}"
```
Comment (jirutka) [inline srcpkgs/dracut/template:77]: Why does it depend on `gummibot`? It’s not needed for booting Linux via EFISTUB.

Comment (ahesford) [reply srcpkgs/dracut/template:77]: The dracut UEFI bundle requires a separate UEFI stub provided (in our case) by gummiboot.
